realm/realm-library/src/main/cpp/util.hpp

```diff
@@ -114,29 +114,36 @@ jclass GetClass(JNIEnv* env, const char* classStr);
 
 // Debug trace
 extern int trace_level;
-extern const char* log_tag;
+extern jclass realmlog_class;
+extern jmethodID log_trace;
+extern jmethodID log_debug;
+extern jmethodID log_info;
+extern jmethodID log_warn;
+extern jmethodID log_error;
+extern jmethodID log_fatal;
+
+
 
 #if TRACE
   #if defined(ANDROID)
     #include <android/log.h>
-    #define LOG_DEBUG ANDROID_LOG_DEBUG
-    #define TR_ENTER() if (trace_level >= 1) { __android_log_print(ANDROID_LOG_DEBUG, log_tag, " --> %s", __FUNCTION__); } else {}
-    #define TR_ENTER_PTR(ptr) if (trace_level >= 1) { __android_log_print(ANDROID_LOG_DEBUG, log_tag, " --> %s %" PRId64, __FUNCTION__, static_cast<int64_t>(ptr)); } else {}
-    #define TR(...) if (trace_level >= 2) { __android_log_print(ANDROID_LOG_DEBUG, log_tag, __VA_ARGS__); } else {}
-    #define TR_ERR(...) if (trace_level >= 0) { __android_log_print(ANDROID_LOG_DEBUG, log_tag, __VA_ARGS__); } else {}
-    #define TR_LEAVE() if (trace_level >= 3) { __android_log_print(ANDROID_LOG_DEBUG, log_tag, " <-- %s", __FUNCTION__); } else {}
+    #define TR_ENTER(env) if (trace_level <= 2) { env->CallStaticVoidMethod(realmlog_class, log_trace, " --> %s", __FUNCTION__); } else {}
+    #define TR_ENTER_PTR(env, ptr) if (trace_level <= 2) { env->CallStaticVoidMethod(realmlog_class, log_trace, " --> %s %", PRId64, __FUNCTION__, static_cast<int64_t>(ptr)); } else {}
+    #define TR(env, ...) if (trace_level <= 2) { env->CallStaticVoidMethod(realmlog_class, log_trace, __VA_ARGS__); } else {}
+    #define TR_ERR(env, ...) if (trace_level <= 6) { env->CallStaticVoidMethod(realmlog_class, log_error, __VA_ARGS__); } else {}
+    #define TR_LEAVE(env) if (trace_level <= 2) { env->CallStaticVoidMethod(realmlog_class, log_trace, " <-- %s", __FUNCTION__); } else {}
   #else // ANDROID
-    #define TR_ENTER()
-    #define TR_ENTER_PTR(ptr)
+    #define TR_ENTER(env)
```

Comment:
I like to have `env` defined in the macro, but that requires lots of changes in jni. Are you sure all of them are changed? Maybe we can keep it like before and change it later.